src/server/proxy_protocol.rs

```diff
@@ -159,71 +202,88 @@ where
         }
     }
 
     #[tracing_attributes::instrument]
-    pub async fn get_session_by_incoming_data_connection(&mut self, connection: &ConnectionTuple) -> Option<SharedSession<S, U>> {
-        let hash = connection.key();
+    pub async fn get_session_by_incoming_data_connection(&mut self, connection: &ProxyConnection) -> Option<SharedSession<S, U>> {
+        let hash: ProxyHashKey = connection.into();
 
         match self.switchboard.get(&hash) {
             Some(session) => session.clone(),
             None => None,
         }
     }
 
-    /// based on source ip of the client, select a free entry
-    /// but initialize it to None
-    // TODO: set a TTL on the hashmap entries
-    #[tracing_attributes::instrument]
+    /// Find the next available port within the specified range (inclusive of the upper limit).
+    /// The reserved port is associated with the source ip of the client and the associated session, using a hashmap
+    ///
+    //#[tracing_attributes::instrument]
     pub async fn reserve_next_free_port(&mut self, session_arc: SharedSession<S, U>) -> Result<u16, ProxyProtocolError> {
-        let rng_length = self.port_range.end - self.port_range.start;
-
-        // change this to a "shuffle" method later on, to make sure we tried all available ports
-        for _ in 1..10 {
-            let random_u32 = {
-                let mut data = [0; 4];
-                getrandom::getrandom(&mut data).expect("Error generating random free port to reserve");
-                u32::from_ne_bytes(data)
-            };
-
-            let port = random_u32 % rng_length as u32 + self.port_range.start as u32;
-            let session = session_arc.lock().await;
-            if let Some(destination) = session.destination {
-                let hash = construct_proxy_hash_key(&destination.ip(), port as u16);
-
-                match &self.try_and_claim(hash.clone(), session_arc.clone()) {
-                    Ok(_) => return Ok(port as u16),
-                    Err(_) => continue,
+        let range_size = self.port_range.end - self.port_range.start;
+
+        let randomized_initial_port = {
+            let mut data = [0; 2];
+            getrandom::getrandom(&mut data).expect("Error generating random free port to reserve");
+            u16::from_ne_bytes(data)
+        };
+
+        // Claims the next available listening port
+        // The search starts at randomized_initial_port.
+        // If a port is already claimed, the loop continues to the next port until an available port is found.
+        // The function returns the first available port it finds or an error if no ports are available.
+        for i in 0..range_size + 1 {
+            let port = self.port_range.start + ((randomized_initial_port + i) % range_size);
+            slog::debug!(self.logger, "Trying if port {} is available", port);
+            let mut session = session_arc.lock().await;
```

Comment:
does it make sense to move this outside the for? Seems like we're dropping and looking continously

Reply:
yeah, good point